src/libraries/System.Private.CoreLib/src/System/Collections/ObjectModel/ReadOnlyCollection.cs

```diff
@@ -229,4 +230,34 @@ void IList.RemoveAt(int index)
             ThrowHelper.ThrowNotSupportedException(ExceptionResource.NotSupported_ReadOnlyCollection);
         }
     }
+
+    /// <summary>
+    /// Provides static methods to create read-only collections and sets.
+    /// </summary>
+    public static class ReadOnlyCollection
+    {
+        /// <summary>
+        /// Creates a new <see cref="ReadOnlyCollection{T}"/> from the specified span of values.
+        /// This method (simplifies collection initialization)[/dotnet/csharp/language-reference/operators/collection-expressions]
+        /// to create a new <see cref="ReadOnlyCollection{T}"/> with the specified values.
+        /// </summary>
+        /// <typeparam name="T">The type of elements in the collection.</typeparam>
+        /// <param name="values">The span of values to include in the collection.</param>
+        /// <returns>A new <see cref="ReadOnlyCollection{T}"/> containing the specified values.</returns>
+        [Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0301:Simplify collection initialization", Justification = "This method simplifies ReadOnlyCollection<T> initialization, so you cannot simplify collection initialization inside it.")]
+        public static ReadOnlyCollection<T> CreateCollection<T>(params ReadOnlySpan<T> values) =>
+            values.IsEmpty ? ReadOnlyCollection<T>.Empty : new ReadOnlyCollection<T>(values.ToArray());
+
+        /// <summary>
+        /// Creates a new <see cref="ReadOnlySet{T}"/> from the specified span of values.
+        /// This method (simplifies collection initialization)[/dotnet/csharp/language-reference/operators/collection-expressions]
+        /// to create a new <see cref="ReadOnlySet{T}"/> with the specified values.
+        /// </summary>
+        /// <typeparam name="T">The type of elements in the collection.</typeparam>
+        /// <param name="values">The span of values to include in the collection.</param>
+        /// <returns>A new <see cref="ReadOnlySet{T}"/> containing the specified values.</returns>
+        [Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0301:Simplify collection initialization", Justification = "This method simplifies ReadOnlySet<T> initialization, so you cannot simplify collection initialization inside it.")]
+        public static ReadOnlySet<T> CreateSet<T>(params ReadOnlySpan<T> values) =>
+            values.IsEmpty ? ReadOnlySet<T>.Empty : new ReadOnlySet<T>((HashSet<T>)[.. values]);
```

Comment:
Using the spread operator on sets doesn't appear to produce optimal code:

https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5nJYBs6AEgIYDOAFgMoCmALgDwAqAfOgGIBOA9gFtaAB3IA7ET3rlgbdgAoAStOAB5MQBsAniPFz0ASzHCAro0oBKZAG9k6e5gDs6ANroAdO8PGzlALr4SAC+BChwJBQ0DCwc3PxComIAsuIm5BpySirq2rpi+kam5lZItkgOZFR0TPqUTOgAvOhi9ADu8oW+7gAy9GIA5ozUFoEVAGZ8UuQAxtTo8qzo9N7eRZZ2DmUVFXWM7gCCwMDy9CMb9iHlDudO6LuBQUA

Consider handrolling the factory?

Reply:
> Using the spread operator on sets doesn't appear to produce optimal code:
> 

Using Count is not entirely appropriate because the enumeration might contain duplicates, and in some cases, there could be many of them. This would result in unnecessary memory allocation.

Additionally, I think that the compiler is capable of optimizing while (enumerator.MoveNext()) loops effectively.

Reply:
I will replace the loop with a manual, but should I use the Count property?

Reply:
> Additionally, I think that the compiler is capable of optimizing while (enumerator.MoveNext()) loops effectively.

The C# compiler shouldn't be emitting use of an enumerator here, just as it doesn't when foreach'ing over a span, but for some reason is. @cston, @RikkiGibson?

Reply:
Looks like a roslyn bug. I see some bits in foreach lowering which are handling the span case and "rewriting as `for`", which are missed in the spread element lowering.

https://github.com/dotnet/roslyn/blob/8e6206417f57fff0a67b3e553f25d1e73899d2bd/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs#L60

https://github.com/dotnet/roslyn/blob/8e6206417f57fff0a67b3e553f25d1e73899d2bd/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs#L1402

Reply:
> I will replace the loop with a manual, but should I use the Count property?

Presumably you mean `.Length`? The input is a ReadOnlySpan.

Reply:
> Yes! I mean the `Length` property. Is it correct to use `Length` here or not? Sometimes a source may contain many duplicates. This will result in unnecessary memory allocation.

Good point. The compiler [doesn't pre-size the HashSet instances either](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQsGIHQAyAlgHYCOGmWAEgIYDOAFgMoCmALgDynsB8cAN1oAbAK6t6cALxwA2njxwAoiVEBbVgCdaAI2Gs8AJVYAHVrXYAKRABo4iGI4CUAXVRA==) so we should do the same here.

> Should I rewrite the loop to a manual for, or will the compiler error be fixed?

A handwritten factory is fine I think, it doesn't fundamentally change the amount of IL being generated.